### ISSUE TITLE
[Tune] Getting poor results with Tune compared to baseline.

### ISSUE THREAD

**Issue by wabala94** [ray-project/ray#4252]
### System information
- Linux Ubuntu 16.04.5
- **Ray installed from**: pip
- **Ray version**: 0.6.3
- **Torch version**: 1.0.1.post2
- **Python version**: 3.6.7
- **Exact command to reproduce**:  

### Describe the problem

Hey everyone, thanks for making a great product. I think it will greatly ease my hyperparameter searches.

The issue I'm running into with my model is that I am consistently getting worse results with Tune as compared to running the model without Tune, even when using the exact same parameters and seed. The FI I get *without* tune at the end of 10 iterations, for example, might be  0.85-0.9. With the same parameters, using tune, I will get a F1 of 0.5-0.6. I am using Tune with the class Trainable API. I have shown my training code below, both in my original train function and with the tune Trainable class. I have checked over the code several times to make sure the loops are right place, but cannot seem to get any reasonable results with Tune. 

### Source code / logs

This is my original training function:

```python 
# set device
torch.backends.cudnn.benchmark = True
device = torch.device('cuda')
torch.backends.cudnn.deterministic = True
torch.manual_seed(42)
np.random.seed(42)

optimizer = define_optimizer(model=model, lr=1e-3, fine_tune=True, differential=False)

def train_model(model, criterion, optimizer, train_dataset, val_dataset,
                train_dataloader, val_dataloader, num_epochs=1,
               inception=False):
    
    for epoch in range(num_epochs):
        
        # go through both training and validation phase
        for phase in ['train', 'val']:
            
            # full labels/prediction list for global assessment of epoch
        
            full_labels = torch.Tensor().long().to(device)
            full_preds = torch.Tensor().long().to(device)
            
            # set model mode
            
            if phase == 'train':
                model.train()
                dataloader = train_dataloader
            else: 
                model.eval()
                dataloader = val_dataloader
                
            running_loss = 0.0
            running_corrects = 0

            # iterate through data
            for batch in dataloader:
                inputs = batch['image']
                labels = batch['label']
                inputs = inputs.to(device)
                labels = labels.to(device)
                
                # zero the gradients
                optimizer.zero_grad()

                # forward
                with torch.set_grad_enabled(phase == 'train'):
                    if inception and phase == 'train': 
                        outputs, _ = model(inputs)
                    else:
                        outputs = model(inputs)
                    _, preds = torch.max(outputs, 1)
                    loss = criterion(outputs, labels)
                    full_labels = torch.cat((full_labels, labels))
                    full_preds = torch.cat((full_preds, preds))

                # backward
                    if phase == 'train':
                        loss.backward()
                        optimizer.step()

                # metrics
                running_loss += loss.item() * inputs.size(0)
                running_corrects += torch.sum(preds == labels.data)


            if phase == 'val':
                epoch_loss = running_loss / len(val_dataset)
                epoch_acc = running_corrects.double() / len(val_dataset)
                epoch_f1 = f1_model_score(full_labels, full_preds)
                
    return model
```

And this is my Tune Trainable source code: 

```python

import ray
import ray.tune as tune
from ray.tune import Trainable

class ray_classifier(Trainable):
    
    def _setup(self, config):
        device = torch.device('cuda')
        torch.manual_seed(42)
        np.seed(42)
        self.train_loader = struct_dict['train_dataloader']
        self.test_loader = struct_dict['val_dataloader']
        self.train_ds = struct_dict['train_dataset']
        self.test_ds = struct_dict['val_dataset']
        self.model = struct_dict['model']
        self.criterion = struct_dict['criterion']
        self.optimizer = define_optimizer(model=struct_dict['model'], 
                                          lr=config['lr'], fine_tune=True, differential=False)
        
        
        
    def _train_iteration(self):
        self.model.train()
        for batch in self.train_loader:
            inputs = batch['image']
            labels = batch['label']
            inputs = inputs.to(device)
            labels = labels.to(device)
            
            self.optimizer.zero_grad()
            
            with torch.set_grad_enabled(True):
                outputs = self.model(inputs)
                loss = self.criterion(outputs, labels)
                loss.backward()
                self.optimizer.step()
                    
    def _test(self):
        
        self.model.eval()
        
        full_labels = torch.Tensor().long().to(device)
        full_preds = torch.Tensor().long().to(device)
        
        running_loss = 0.0
        running_corrects = 0
        
        for batch in self.test_loader:
            inputs = batch['image']
            labels = batch['label']
            inputs = inputs.to(device)
            labels = labels.to(device)
            with torch.set_grad_enabled(False):
                outputs = model(inputs)
                _, preds = torch.max(outputs, 1)
                loss = self.criterion(outputs, labels)
                running_loss += loss.item() * inputs.size(0)
                running_corrects += torch.sum(preds == labels.data)
                full_labels = torch.cat((full_labels, labels))
                full_preds = torch.cat((full_preds, preds))
            
        epoch_loss = running_loss / len(self.test_ds)
        epoch_acc = running_corrects.double() / len(self.test_ds)
        epoch_f1 = f1_model_score(full_labels, full_preds)
        
        return {
            'mean_loss': epoch_loss,
            'mean_accuracy': epoch_acc.item(),
            'f1': epoch_f1
        }
             
    def _train(self):
        self._train_iteration()
        return self._test()
    
    def _save(self, checkpoint_dir=CKPT_DIR):
        checkpoint_path = os.path.join(checkpoint_dir, 'model.pth')
        torch.save(self.model.state_dict(), checkpoint_path)
        return checkpoint_path
    
    def _restore(self, checkpoint_path):
        return self.model.load_state_dict(checkpoint_path)

```

And this is my experiment setup:

```python

# Employ hyperopt search algorithm

from hyperopt import hp

# define search space (NOTE: Ensuring that we only use one learning rate to test ray function)

space = {
    'lr': hp.choice('lr', [1e-3, 1e-3])
}

# define algorithm with above details

algo = tune.suggest.HyperOptSearch(space, max_concurrent=4,
                                  reward_attr='f1')

# resources per trial

rpt = {'cpu': 1, 'gpu': 1}

# make experiment

exp_2 = tune.Experiment(name='exp_2', run=ray_classifier, stop={'f1': 0.95},
                      resources_per_trial=rpt, num_samples=5,
                     checkpoint_freq=0, checkpoint_at_end=True)

# Define scheduler

scheduler = ray.tune.schedulers.AsyncHyperBandScheduler(time_attr='training_iteration',
                                                        reward_attr='f1',
                                                        grace_period=5,
                                                        reduction_factor=3,
                                                        brackets=3,
                                                        max_t=20)

# initialize ray

ray.init(ignore_reinit_error=True)

tune.run_experiments(experiments=exp_2, verbose=2,
                     search_alg=algo,
                     scheduler=scheduler,
                     resume=False,
                     raise_on_failed_trial=True)

```

Anything you can suggest to lead me in the right direction would be incredibly helpful. Thank you!

**Comment by richardliaw** [comment]
can you try running to check the model training without Tune?

```
trainable = ray_classifier(cfg)
results = [] 
for i in range(10):
    res = trainable.train()
    results += [res]

# analyze results here
```

This would help isolate what possible issues might exist from porting the code.

**Comment by wabala94** [comment]
Hi Richard,

Thanks for the reply.

I've posted the results of running these tests below. Noticeably, the loss does seem to be decreasing at each iteration. In that case, the results might be coming from a class imbalance in my data leading to erratic f1 values (I do have a class with ~1/10 the size of my largest class).  What do you think?

```python

dummy_config = {
    'lr': 1e-3
}
trainable = ray_classifier(dummy_config)

results = []
for i in range(10):
    res = trainable.train()
    results += [res]

```
```
[{'mean_loss': 1.9169575028949313,
  'mean_accuracy': 0.8111111111111111,
  'f1': 0.3717679944095038,
  'done': False,
  'timesteps_total': None,
  'episodes_total': None,
  'neg_mean_loss': -1.9169575028949313,
  'experiment_id': '5db0dc87890c4d3980cc57e71a6720d7',
  'date': '2019-03-04_19-01-49',
  'timestamp': 1551754909,
  'training_iteration': 1,
  'time_this_iter_s': 2.557607889175415,
  'time_total_s': 2.557607889175415,
  'pid': 27888,
  'hostname': 'u09082017000',
  'node_ip': '171.65.34.41',
  'config': {'lr': 0.001},
  'time_since_restore': 2.557607889175415,
  'timesteps_since_restore': 0,
  'iterations_since_restore': 1},
 

{'mean_loss': 1.4514876100752088,
  'mean_accuracy': 0.35555555555555557,
  'f1': 0.24112388250319286,
  'done': False,
  'timesteps_total': None,
  'episodes_total': None,
  'neg_mean_loss': -1.4514876100752088,
  'experiment_id': '5db0dc87890c4d3980cc57e71a6720d7',
  'date': '2019-03-04_19-01-52',
  'timestamp': 1551754912,
  'training_iteration': 2,
  'time_this_iter_s': 2.451181411743164,
  'time_total_s': 5.008789300918579,
  'pid': 27888,
  'hostname': 'u09082017000',
  'node_ip': '171.65.34.41',
  'config': {'lr': 0.001},
  'time_since_restore': 5.008789300918579,
  'timesteps_since_restore': 0,
  'iterations_since_restore': 2},


 {'mean_loss': 0.24127853181627062,
  'mean_accuracy': 0.9111111111111111,
  'f1': 0.57183908045977,
  'done': False,
  'timesteps_total': None,
  'episodes_total': None,
  'neg_mean_loss': -0.24127853181627062,
  'experiment_id': '5db0dc87890c4d3980cc57e71a6720d7',
  'date': '2019-03-04_19-01-54',
  'timestamp': 1551754914,
  'training_iteration': 3,
  'time_this_iter_s': 2.3780128955841064,
  'time_total_s': 7.3868021965026855,
  'pid': 27888,
  'hostname': 'u09082017000',
  'node_ip': '171.65.34.41',
  'config': {'lr': 0.001},
  'time_since_restore': 7.3868021965026855,
  'timesteps_since_restore': 0,
  'iterations_since_restore': 3},


 {'mean_loss': 0.1704251143667433,
  'mean_accuracy': 0.9222222222222223,
  'f1': 0.5906103286384976,
  'done': False,
  'timesteps_total': None,
  'episodes_total': None,
  'neg_mean_loss': -0.1704251143667433,
  'experiment_id': '5db0dc87890c4d3980cc57e71a6720d7',
  'date': '2019-03-04_19-01-56',
  'timestamp': 1551754916,
  'training_iteration': 4,
  'time_this_iter_s': 2.3431906700134277,
  'time_total_s': 9.729992866516113,
  'pid': 27888,
  'hostname': 'u09082017000',
  'node_ip': '171.65.34.41',
  'config': {'lr': 0.001},
  'time_since_restore': 9.729992866516113,
  'timesteps_since_restore': 0,
  'iterations_since_restore': 4},


 {'mean_loss': 0.3585824728012085,
  'mean_accuracy': 0.888888888888889,
  'f1': 0.7745768947755703,
  'done': False,
  'timesteps_total': None,
  'episodes_total': None,
  'neg_mean_loss': -0.3585824728012085,
  'experiment_id': '5db0dc87890c4d3980cc57e71a6720d7',
  'date': '2019-03-04_19-01-59',
  'timestamp': 1551754919,
  'training_iteration': 5,
  'time_this_iter_s': 2.3353118896484375,
  'time_total_s': 12.06530475616455,
  'pid': 27888,
  'hostname': 'u09082017000',
  'node_ip': '171.65.34.41',
  'config': {'lr': 0.001},
  'time_since_restore': 12.06530475616455,
  'timesteps_since_restore': 0,
  'iterations_since_restore': 5},


 {'mean_loss': 0.3680287235312992,
  'mean_accuracy': 0.9333333333333333,
  'f1': 0.5972701149425287,
  'done': False,
  'timesteps_total': None,
  'episodes_total': None,
  'neg_mean_loss': -0.3680287235312992,
  'experiment_id': '5db0dc87890c4d3980cc57e71a6720d7',
  'date': '2019-03-04_19-02-01',
  'timestamp': 1551754921,
  'training_iteration': 6,
  'time_this_iter_s': 2.36641263961792,
  'time_total_s': 14.43171739578247,
  'pid': 27888,
  'hostname': 'u09082017000',
  'node_ip': '171.65.34.41',
  'config': {'lr': 0.001},
  'time_since_restore': 14.43171739578247,
  'timesteps_since_restore': 0,
  'iterations_since_restore': 6},


 {'mean_loss': 0.2986372987429301,
  'mean_accuracy': 0.9333333333333333,
  'f1': 0.8505668934240364,
  'done': False,
  'timesteps_total': None,
  'episodes_total': None,
  'neg_mean_loss': -0.2986372987429301,
  'experiment_id': '5db0dc87890c4d3980cc57e71a6720d7',
  'date': '2019-03-04_19-02-03',
  'timestamp': 1551754923,
  'training_iteration': 7,
  'time_this_iter_s': 2.3821797370910645,
  'time_total_s': 16.813897132873535,
  'pid': 27888,
  'hostname': 'u09082017000',
  'node_ip': '171.65.34.41',
  'config': {'lr': 0.001},
  'time_since_restore': 16.813897132873535,
  'timesteps_since_restore': 0,
  'iterations_since_restore': 7},


 {'mean_loss': 0.27530544069078233,
  'mean_accuracy': 0.8777777777777778,
  'f1': 0.5576279292204687,
  'done': False,
  'timesteps_total': None,
  'episodes_total': None,
  'neg_mean_loss': -0.27530544069078233,
  'experiment_id': '5db0dc87890c4d3980cc57e71a6720d7',
  'date': '2019-03-04_19-02-06',
  'timestamp': 1551754926,
  'training_iteration': 8,
  'time_this_iter_s': 2.4266128540039062,
  'time_total_s': 19.24050998687744,
  'pid': 27888,
  'hostname': 'u09082017000',
  'node_ip': '171.65.34.41',
  'config': {'lr': 0.001},
  'time_since_restore': 19.24050998687744,
  'timesteps_since_restore': 0,
  'iterations_since_restore': 8},


 {'mean_loss': 0.4061767684088813,
  'mean_accuracy': 0.8222222222222223,
  'f1': 0.7207296849087893,
  'done': False,
  'timesteps_total': None,
  'episodes_total': None,
  'neg_mean_loss': -0.4061767684088813,
  'experiment_id': '5db0dc87890c4d3980cc57e71a6720d7',
  'date': '2019-03-04_19-02-08',
  'timestamp': 1551754928,
  'training_iteration': 9,
  'time_this_iter_s': 2.375826358795166,
  'time_total_s': 21.616336345672607,
  'pid': 27888,
  'hostname': 'u09082017000',
  'node_ip': '171.65.34.41',
  'config': {'lr': 0.001},
  'time_since_restore': 21.616336345672607,
  'timesteps_since_restore': 0,
  'iterations_since_restore': 9},

 {'mean_loss': 0.39441212945514253,
  'mean_accuracy': 0.8666666666666667,
  'f1': 0.5280405868641164,
  'done': False,
  'timesteps_total': None,
  'episodes_total': None,
  'neg_mean_loss': -0.39441212945514253,
  'experiment_id': '5db0dc87890c4d3980cc57e71a6720d7',
  'date': '2019-03-04_19-02-11',
  'timestamp': 1551754931,
  'training_iteration': 10,
  'time_this_iter_s': 2.403423309326172,
  'time_total_s': 24.01975965499878,
  'pid': 27888,
  'hostname': 'u09082017000',
  'node_ip': '171.65.34.41',
  'config': {'lr': 0.001},
  'time_since_restore': 24.01975965499878,
  'timesteps_since_restore': 0,
  'iterations_since_restore': 10}]
```

**Comment by richardliaw** [comment]
Hm, well specifically regarding the Tune porting issue, technically if you
set the epochs to 1 and seed properly, the output of both 1 training
iteration and 1 loop iteration of your prior code should be exactly the
same. Can you check if this is the case? Otherwise, there might be some odd
typo.

**Comment by wabala94** [comment]
Ah, I see what you're getting at now. I looked back to make sure I was using the torch reproducibility guidelines and I am getting the same result with 1 iteration! In that case, the port is probably fine.

**Comment by richardliaw** [comment]
Awesome; feel free to close this (also, feel free to contribute what you’re
doing as an example!)

**Comment by wabala94** [comment]
Thank you!
